src/segment-loader.js

```diff
@@ -1178,6 +1221,187 @@ export default class SegmentLoader extends videojs.EventTarget {
+      maxReverseBuffer = (reverseBuffer / configTotalBuffer) * upperTimeLimit;
+    }
+
+    maxForwardBuffer = Math.min(forwardBuffer, maxForwardBuffer);
+    maxReverseBuffer = Math.min(reverseBuffer, maxReverseBuffer);
+
+    loggerMsg = '\n\tbytes in buffer: ' + bytesInBuffer +
+      ', seconds in buffer: ' + secondsInBuffer +
+      '\n\tforward buffer: ' + forwardBuffer + ' => ' + maxForwardBuffer +
+      '\n\treverse buffer: ' + reverseBuffer + ' => ' + maxReverseBuffer;
+
+    if (setByteLimit) {
+      loggerMsg += '\n\tbyte limit: ' + this.maxBytes_ + ' => ' + bytesInBuffer;
+      this.maxBytes_ = Math.min(this.maxBytes_, bytesInBuffer);
+    }
+    this.logger_('adjustBuffers_', loggerMsg);
+
+    this.goalBufferLength_(maxForwardBuffer);
+    this.backBufferLength_(maxReverseBuffer);
+  }
+
+  /**
+   * handler to run when a call to appendBuffer fails because the source
+   * buffer can't hold any more information
+   *
+   * @private
+   *
+   * @param {Object} event
+   * @param {Function} event.done The callback to run when the append finally succeeds
+   * @param {Number=} event.safeRemovePoint The most recent GOP start time, if we have it
+   * @param {Object} event.segment The Uint8Array for the segment
+   * @param {Object} event.target The SourceBuffer that caused the error
+   */
+  handleBufferMaxed_(event) {
+    let bytesTried = event.segment.byteLength;
+    let segmentBytes = event.segment;
+    let sourceBuffer = event.target;
+    let safeRemovePoint = event.safeRemovePoint;
+    let pendingAppend = true;
+    let callback = event.done;
+
+    // if we fail on the first request, it's probably a glitch, and
+    // we should just slice the segment up into small bits and append
+    if (this.mediaRequests === 1) {
+      let slices = [];
+      let start = 0;
+      let end = 0;
+      let sliceAppend = () => {
+        if (slices.length) {
+          sourceBuffer.appendBuffer(slices.shift());
+        } else {
+          sourceBuffer.removeEventListener('updateend', sliceAppend);
+          callback();
+        }
+      };
+
+      while (start < bytesTried) {
+        end += 5000000;
+        if (end > bytesTried) {
+          end = bytesTried;
+        }
+        slices.push(segmentBytes.slice(start, end));
+        start += 5000000;
+      }
+      sourceBuffer.addEventListener('updateend', sliceAppend);
+      sourceBuffer.appendBuffer(slices.shift());
+      return;
+    }
+
+    let bytesTrimmable;
+    let removeToTime;
+    let timeupdateHandler;
+    let tryAppend = () => {
+      if (!pendingAppend) {
+        return;
+      }
+      this.hls_.tech_.off('timeupdate', timeupdateHandler);
+      this.logger_('Appended deferred segment with ' +
+        (this.buffered_().end(0) - this.currentTime_()) + ' to spare...');
+      // we don't want this to be called again
+      sourceBuffer.removeEventListener('updateend', tryAppend);
+      try {
+        sourceBuffer.appendBuffer(segmentBytes);
```

Comment:
This is inside the `tryAppend` function.

Reply:
- Append blocks any future requests until its `updateend` handler is called
- When the `QuotaExceededError` happens, we unblock the source updater (by nulling out `pendingCallback_`, but nothing will still happen automatically since the `updateend` handler has not fired
- In `handleBuffersMaxed`, we never call an `append` before a `remove`
- Therefore we'll never have `tryAppend` fire from an `append`, it will only be fired from a `remove`